---
title: 'CA1031: не перехватывайте типы общих исключений'
ms.date: 11/04/2016
ms.topic: reference
f1_keywords:
- CA1031
- DoNotCatchGeneralExceptionTypes
helpviewer_keywords:
- CA1031
- DoNotCatchGeneralExceptionTypes
ms.assetid: cbc283ae-2a46-4ec0-940e-85aa189b118f
author: jillre
ms.author: jillfra
manager: jillfra
dev_langs:
- CPP
- CSharp
- VB
ms.workload:
- multiple
ms.openlocfilehash: 8f4eba9dd06b63aaa61c0893e8dc3e37b649777e
ms.sourcegitcommit: a8e8f4bd5d508da34bbe9f2d4d9fa94da0539de0
ms.translationtype: MT
ms.contentlocale: ru-RU
ms.lasthandoff: 10/19/2019
ms.locfileid: "72616960"
---
# <a name="ca1031-do-not-catch-general-exception-types"></a>CA1031: не перехватывайте типы общих исключений

|||
|-|-|
|TypeName|DoNotCatchGeneralExceptionTypes|
|CheckId|CA1031|
|Категория|Microsoft. Design|
|Критическое изменение|Не критическое|

## <a name="cause"></a>Причина
В инструкции `catch` перехватывается общее исключение, например <xref:System.Exception?displayProperty=fullName> или <xref:System.SystemException?displayProperty=fullName>, или используется общее предложение catch, такое как `catch()`.

## <a name="rule-description"></a>Описание правила
Общие исключения не должны перехватываться.

## <a name="how-to-fix-violations"></a>Устранение нарушений
Чтобы устранить нарушение этого правила, перехватите более конкретное исключение или повторно создайте общее исключение в качестве последней инструкции в блоке `catch`.

## <a name="when-to-suppress-warnings"></a>Когда следует подавлять предупреждения
Для этого правила отключать вывод предупреждений не следует. Перехват общих типов исключений может скрывать проблемы времени выполнения от пользователя библиотеки и может усложнить отладку.

> [!NOTE]
> Начиная с .NET Framework 4, общеязыковая среда выполнения (CLR) больше не доставляет исключения поврежденного состояния, происходящие в операционной системе, и управляемый код, например нарушения прав доступа в [!INCLUDE[TLA#tla_mswin](../code-quality/includes/tlasharptla_mswin_md.md)], которые должны обрабатываться управляемым кодом. Если вы хотите скомпилировать приложение в .NET Framework 4 или более поздней версии и поддерживать обработку исключений поврежденного состояния, можно применить атрибут <xref:System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute> к методу, обрабатывающему исключение поврежденного состояния.

## <a name="example"></a>Пример
В следующем примере показан тип, нарушающий это правило, и тип, который правильно реализует блок `catch`.

[!code-cpp[FxCop.Design.ExceptionAndSystemException#1](../code-quality/codesnippet/CPP/ca1031-do-not-catch-general-exception-types_1.cpp)]
[!code-vb[FxCop.Design.ExceptionAndSystemException#1](../code-quality/codesnippet/VisualBasic/ca1031-do-not-catch-general-exception-types_1.vb)]
[!code-csharp[FxCop.Design.ExceptionAndSystemException#1](../code-quality/codesnippet/CSharp/ca1031-do-not-catch-general-exception-types_1.cs)]

## <a name="related-rules"></a>Связанные правила
[CA2200: следует повторно вызывать исключение для сохранения сведений о стеке](../code-quality/ca2200.md)
